Libraries/D3D12RaytracingFallback/src/TreeletReorder.hlsl

```diff
@@ -56,210 +45,289 @@ AABB CombineAABB(AABB aabb0, AABB aabb1)
-            for (int j = numInternalTreeletNodes - 1; j >= 0; j--)
+    // Start from the back. This is optimizing since the previous traversal went from
+    // top-down, the reverse order is guaranteed to be bottom-up
+    [unroll]
+    for (int j = NumInternalTreeletNodes - 1; j >= 0; j--)
+    {
+        uint internalNodeIndex = internalNodes[j];
+        AABB leftAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].LeftChildIndex];
+        AABB rightAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].RightChildIndex];
+        AABBBuffer[internalNodeIndex] = CombineAABB(leftAABB, rightAABB);
+    }
+}
+
+void MoveToParent(in uint threadId)
+{
+    if (threadId == 0)
+    {
+        if (nodeIndex == RootNodeIndex)
+        {
+            finished = true;
+        }
+        else
+        {
+            uint parentNodeIndex = hierarchyBuffer[nodeIndex].ParentIndex;
+
+            uint ourNumTriangles = NumTrianglesBuffer.Load(nodeIndex * SizeOfUINT32);
+            uint numTrianglesFromOtherNode = 0;
+            NumTrianglesBuffer.InterlockedAdd(parentNodeIndex * SizeOfUINT32, ourNumTriangles, numTrianglesFromOtherNode);
+
+            // Wait for sibling in tree
+            if (numTrianglesFromOtherNode == 0)
             {
-                uint internalNodeIndex = internalNodes[j];
-                AABB leftAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].LeftChildIndex];
-                AABB rightAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].RightChildIndex];
-                AABBBuffer[internalNodeIndex] = CombineAABB(leftAABB, rightAABB);
+                finished = true;
+            }
+            else
+            {
+                // Build parents bounding box
+                AABB leftAABB = AABBBuffer[hierarchyBuffer[parentNodeIndex].LeftChildIndex];
+                AABB rightAABB = AABBBuffer[hierarchyBuffer[parentNodeIndex].RightChildIndex];
+                AABBBuffer[parentNodeIndex] = CombineAABB(leftAABB, rightAABB);
+
+                nodeIndex = parentNodeIndex;
             }
         }
+    }
 
-        AABBBuffer[nodeIndex] = nodeAABB;
-        DeviceMemoryBarrier(); // Ensure AABBs have been written out and are visible to all waves
+    GroupMemoryBarrierWithGroupSync();
+}
+
+[numthreads(NumThreadsInGroup, 1, 1)]
+void main(uint3 Gid : SV_GroupID, uint3 GTid : SV_GroupThreadID)
+{
+    if (GTid.x == 0)
+    {
+        nodeIndex = BaseTreeletsIndexBuffer[Gid.x];
+        finished = false;
+    }
+
+    GroupMemoryBarrierWithGroupSync();
+
+    const uint NumberOfAABBs = GetNumInternalNodes(Constants.NumberOfElements) + Constants.NumberOfElements;
+
+    if (nodeIndex >= NumberOfAABBs)
+    {
+        return;
+    }
+
+    do
+    {
+        FormTreelet(GTid.x);
+
+        FindOptimalPartitions(GTid.x);
 
-        if (nodeIndex == rootNodeIndex) return;
+        ReformTree(GTid.x);
 
-        uint parentNodeIndex = hierarchyBuffer[nodeIndex].ParentIndex;
+        MoveToParent(GTid.x);
```

Comment:
Very nitpicky, but I'd rename this "TraverseToParent". Move can have the semantics of things like std::move where you might be assigning something to the parent rather than traversal which is less overloaded